File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        long n= prices.size();
        if(n==1) return 0;
        vector<long> maxPrices(n,INT_MIN);
        for(auto i=n-2;i>=0;--i){
            maxPrices[i] = max((long)prices[i+1],maxPrices[i+1]);
        }
        
        long maxProfit = 0;
        for(long i=0;i<n;i++){
            maxProfit = max(maxProfit,maxPrices[i]-prices[i]);
        }
        
        return maxProfit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minBefore = inf
        maxProfit = 0
        for i in range(len(prices)):
            cur = prices[i]
            maxProfit = max(maxProfit,cur-minBefore)
            minBefore = min(minBefore,cur)
        return maxProfit